qualtran/bloqs/state_preparation/controlled_state_preparation.py

```diff
@@ -0,0 +1,484 @@
+            self.state_bitsizes, self.rot_reg_bitsizes, tuple(rom_vals)
+        )
+        control, target_state, rot_ancilla, rot_reg, phase_gradient = bb.add(
+            ctrl_rot,
+            prepare_control=control,
+            selection=target_state,
+            qubit=rot_ancilla,
+            rot_reg=rot_reg,
+            phase_gradient=phase_gradient,
+        )
+        bb.add(OneEffect(), q=rot_ancilla)
+        return control, target_state, rot_reg, phase_gradient
+
+    def _get_phase_rom_values(self, amplitude_rom_vals):
+        """As we are using the equivalent to controlled Z to do the rotations instead of Rz, there
+        is a phase offset for each coefficient that has to be corrected. This offset is half of the
+        turn angle applied, and is added to the phase for each coefficient.
+        """
+        offset_angles = [0] * (2**self.state_bitsizes)
+        for i in range(self.state_bitsizes):
+            for j in range(2**i):
+                item_range = 2 ** (self.state_bitsizes - i)
+                # if the rom has value 0 the formula gives 180, when it should be 0
+                if amplitude_rom_vals[i][j] == 0:
+                    offset = 0
+                else:
+                    offset = np.pi * amplitude_rom_vals[i][j] / (2**self.rot_reg_bitsizes)
+                for k in range(item_range * j, item_range * (j + 1)):
+                    offset_angles[k] += offset
+        # if the matrix is the adjoint, the angles have to be undone, thus just load -theta
+        if self.uncompute:
+            angles = [offset - np.angle(c) for c, offset in zip(self.state, offset_angles)]
+        else:
+            angles = [np.angle(c) - offset for c, offset in zip(self.state, offset_angles)]
+        rom_values = [RotationTree.angle_to_rom_value(a, self.rot_reg_bitsizes) for a in angles]
+        return rom_values
+
+
+@bloq_example
+def _controlled_state_prep() -> ControlledStatePreparationUsingRotations:
+    state_coefs = (
+        (-0.42677669529663675 - 0.1767766952966366j),
+        (0.17677669529663664 - 0.4267766952966367j),
+        (0.17677669529663675 - 0.1767766952966368j),
+        (0.07322330470336305 - 0.07322330470336309j),
+        (0.4267766952966366 - 0.17677669529663692j),
+        (0.42677669529663664 + 0.17677669529663675j),
+        (0.0732233047033631 + 0.17677669529663678j),
+        (-0.07322330470336308 - 0.17677669529663678j),
+    )
+    controlled_state_prep = ControlledStatePreparationUsingRotations(
+        state_bitsizes=2, rot_reg_bitsizes=4, state=state_coefs
+    )
+    return controlled_state_prep
+
+
+_CONTROLLED_STATE_PREP_DOC = BloqDocSpec(
+    bloq_cls=ControlledStatePreparationUsingRotations,
+    import_line='from qualtran.bloqs.state_preparation.controlled_state_preparation import ControlledStatePreparationUsingRotations',
+    examples=(_controlled_state_prep,),
+)
+
+
+@attrs.frozen
+class ControlledQROMRotateQubit(Bloq):
```

Comment:
I think it would be nice to add a more general version of it in the same family as `ProgrammableRotationGateArray` (for example: support arbitrary number of controls, including 0) but we can do that as a follow up PR. For now, it'd be sufficient to open an issue and track improvements to this important primitive.

Reply:
I created the [issue](https://github.com/quantumlib/Qualtran/issues/676) but can't assign it to myself.